pkg/operator/operator_test.go

```diff
@@ -228,11 +240,59 @@ func TestOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
 				Platform: v1.AWSPlatformType,
 			},
 		}
-
 		stopCh := make(<-chan struct{})
-		optr := newFakeOperator(nil, []runtime.Object{tc.featureGate, infra}, stopCh)
-		go optr.Run(2, stopCh)
+		optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
+		err := initFeatureGateInformer([]runtime.Object{tc.featureGate}, optr)
+		if err != nil {
+			t.Errorf("Failed to set featuregate %v : %v", tc.featureGate, err)
+		}
+		optr.sync("test-key")
+		if err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+			d, err := optr.deployLister.Deployments(targetNamespace).Get(deploymentName)
+			if err != nil {
+				t.Logf("Failed to get %q deployment: %v", deploymentName, err)
+				return false, nil
+			}
+			if deploymentHasContainer(d, hcControllerName) != tc.expectedMachineHealthCheckController {
+				t.Logf("Expected deploymentHasContainer for %q container to be %t", hcControllerName, tc.expectedMachineHealthCheckController)
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			t.Errorf("Failed to verify %q deployment", deploymentName)
+		}
+	}
+}
 
+func TestEnableDisableOperatorSyncClusterAPIControllerHealthCheckController(t *testing.T) {
+	tests := []struct {
+		featureGate                          *v1.FeatureGate
+		expectedMachineHealthCheckController bool
+	}{{
+		featureGate:                          newFeatureGate(v1.TechPreviewNoUpgrade),
+		expectedMachineHealthCheckController: true,
+	}, {
+		featureGate:                          newFeatureGate(v1.Default),
+		expectedMachineHealthCheckController: false,
+	}}
+	infra := &v1.Infrastructure{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster",
+		},
+		Status: v1.InfrastructureStatus{
+			Platform: v1.AWSPlatformType,
+		},
+	}
+
+	stopCh := make(<-chan struct{})
+	optr := newFakeOperator(nil, []runtime.Object{infra}, stopCh)
+
+	for _, tc := range tests {
```

Comment:
Given the unit test will always have two steps (first enable fg, then disable fg), there is no need to iterate through both tests. It's sufficient to initialize `newFakeOperator` as it was before, remove `initFeatureGateInformer` and just call `optr.sync` two times. After the first call validating the fg is enable, after the second call validating it is not.

Reply:
>  remove initFeatureGateInformer

This is not possible because modifying the `FeatureGate` object which is passed to the fake clientset, will not work.  Eventually it makes a copy of the object and therefor any changes to the object will not be reflected back in the object returned by informer in response to the listing operation.

Updated the PR to address other parts. PTAL!

Reply:
```

Reply:
@vikaschoudhary16 also please improve the messages to be more human readable

Reply:
updated, thanks!